include/fuzzuf/coverage/shm_coverage.hpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+/**
+ * @file shm_coverage.hpp
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ */
+
+#ifndef FUZZUF_INCLUDE_COVERAGE_SHM_COVERAGE_HPP
+#define FUZZUF_INCLUDE_COVERAGE_SHM_COVERAGE_HPP
+
+#include "fuzzuf/feedback/inplace_memory_feedback.hpp"
+#include "fuzzuf/logger/logger.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+/**
+ * @class ShmCoverage
+ * @brief Base class that represents coverage stored in shared memory.
+ */
+class ShmCoverage {
+public:
+  static constexpr int INVALID_SHMID = -1;
+
+  u8 *trace_bits;
+
+  ShmCoverage(u32 map_size)
+  : trace_bits(nullptr),
+    map_size(map_size),
+    shmid(INVALID_SHMID) {}
+
+  ~ShmCoverage() {
+    Erase();
+  }
+
+  ShmCoverage(const ShmCoverage&) = delete;
+  ShmCoverage(ShmCoverage &&) = delete;
+  ShmCoverage &operator=(const ShmCoverage&) = delete;
+  ShmCoverage &operator=(ShmCoverage&&) = delete;
+  ShmCoverage() = delete;
+
+  void Setup(void) {
+    if (map_size == 0) return;
+    shmid = shmget(IPC_PRIVATE, map_size, IPC_CREAT | IPC_EXCL | 0600);
+    if (shmid < 0) ERROR("shmget() failed");
+
+    trace_bits = (u8 *)shmat(shmid, nullptr, 0);
+    if (trace_bits == (u8 *)-1) ERROR("shmat() failed");
+  }
+
+  void Reset(void) {
+    if (map_size == 0) return;
+    if (trace_bits != nullptr) {
+      std::memset(trace_bits, 0, map_size);
+    }
+
+    MEM_BARRIER();
+  }
+
+  void Erase(void) {
+    if (map_size == 0) return;
+    if (trace_bits != nullptr) {
+      if (shmdt(trace_bits) == -1) ERROR("shmdt() failed");
+      trace_bits = nullptr;
+    }
+    if (shmid != INVALID_SHMID) {
+      if (shmctl(shmid, IPC_RMID, 0) == -1) ERROR("shmctl() failed");
+      shmid = INVALID_SHMID;
+    }
+  }
+
+  void SetupEnvironmentVariable(const char *shm_env_var) {
+    if (map_size > 0 && shmid != INVALID_SHMID) {
+      std::string shmstr = std::to_string(shmid);
+      setenv(shm_env_var, shmstr.c_str(), 1);
+    } else {
+      unsetenv(shm_env_var);
+    }
+  }
+
+  virtual u32 GetMapSize(void) {
+    return map_size;
+  }
+
+  virtual int GetShmID(void) {
+    return shmid;
+  }
+
+  InplaceMemoryFeedback GetFeedback(std::shared_ptr<u8> lock) {
+    return InplaceMemoryFeedback(trace_bits, map_size, lock);
+  }
```

Comment:
I think it is `ShmCoverage` that should have `lock`. If some executor wrongly passes different locks to the same instance of this class, then the returned feedbacks have different locks and hence we can overwrite the shared memory although it may be referenced to in multiple locations.

Reply:
I changed to use each coverage attacher has `lock`, and the executor refers to the locks in commit 
77a8053.